### DISCUSSION TITLE
A way to add Configuration needed for cart-and-checkout-validation-api functions

### DISCUSSION THREAD

**Discussion by raj-kumar94** [Shopify/function-examples#206] [Feedback: Cart and Checkout Validation API]
Hi, currently is there no way to store and retrieve the configuration for the function?

Eg: The example function in [this guide](https://shopify.dev/docs/apps/checkout/validation/cart-checkout) uses `orderSubtotal` value of $1000. This value will differ from merchant to merchant. So we need a way to store this in a metafield and retrieve it in the function.

I tried the following

```
mutation {
  metafieldsSet(metafields: [
    {
      ownerId: "gid://shopify/CheckoutServerExtension/33074"
      namespace: "checkout-customization"
      key: "function-configuration"
      value: "{ \"foo\": \"bar\"}"
      type: "json"
    }
  ]) {
    metafields {
      id
    }
    userErrors {
      message
    }
  }
}
```

But got "invalid id" error

```
{
  "data": {
    "metafieldsSet": null
  },
  "errors": [
    {
      "message": "invalid id",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "metafieldsSet"
      ]
    }
  ],
  "extensions": {
    "cost": {
      "requestedQueryCost": 10,
      "actualQueryCost": 1,
      "throttleStatus": {
        "maximumAvailable": 1000,
        "currentlyAvailable": 999,
        "restoreRate": 50
      }
    }
  }
}
```

If this is already in pipeline, that would be awesome!!
Excited to explore the possibilities with this API 🤩

**Comment by realpratiknikam** [comment]
Without merchant configuration, cart validations are kind of use-less so support for merchant config would be greatly appreciated!

**Comment by pateketu** [comment]
For the use cases, I was looking at, I need to supply some settings/config data which will differ from merchant to merchant, otherwise, validation functions are pretty useless, can only do some basic logic, and I don't want to litter metafields of each product with duplicate information, another option is to allow shop metafields access in function then app devs can populate shop metafield with configuration data for each merchangt

**Comment by gil--** [comment]
Need to block checkout based on zipcodes but those differ merchant to merchant and merchants want to update the list continuously which is not possible without a metafield on the function. 

Noticed metafield is not supported for this function which seems to be unique to the validation function - https://shopify.dev/docs/api/functions/reference/cart-checkout-validation/graphql/common-objects/metafield

**Comment by aledalgrande** [comment]
Hi folks, thanks for the feedback. Support for configuration via metafields is coming soon.

**Reply by gil--** [reply]
This is really exciting to hear! Thank you team :)

**Reply by axtricssolutions** [reply]
Hey, Any update on the Metafields for them to be available in Validation Functions API?

**Reply by axtricssolutions** [reply]
@aledalgrande Just curious when it will be available?

**Reply by The1987** [reply]
@aledalgrande - any update on this?

**Reply by AndreaMorone** [reply]
We found a solution for that, we are saving configuration on Shop metafield and we retrieve it in the function input. Having a dedicated metafield like how it is for payment customizaitons would be better, but it's working fine. We also had also a solution where we were saving it on Localization -> Market -> Metafield.

**Reply by The1987** [reply]
@AndreaMorone thank you. any good examples out there in github or documentation you'd recommend? I'll have to read up on it. Just getting started with checkout extensions (already built other apps prior and have used TAE).

**Reply by AndreaMorone** [reply]
Take a look at the example apps for functions, choose one to understand the logics, and then just save the the config into a shop metafield that you can retrieve in the function input 😊

**Reply by aledalgrande** [reply]
Thanks for the patience peeps, this is in the works.

**Comment by WingPlex** [comment]
The metafields are working but still variables are not working. [here](https://github.com/Shopify/function-examples/issues/377)

**Comment by GMatrixGames** [comment]
Is there any update on this?

**Reply by aledalgrande** [reply]
If you meant adding configuration to the function that is possible since 2024-04: https://shopify.dev/docs/api/functions/reference/cart-checkout-validation/graphql/common-objects/validation?api%5Bversion%5D=2024-04

Tutorials here: https://shopify.dev/docs/apps/build/checkout/cart-checkout-validation

**Reply by GMatrixGames** [reply]
Actually, I see that what's implemented is not what I'm looking for. I got confused I guess, I thought this was for Input variables.
Unless I am missing something, I can't find info on those either. Maybe you could point me in the right direction?

**Reply by aledalgrande** [reply]
Have you already tried this? https://shopify.dev/docs/apps/build/functions/input-output/use-variables-input-queries

**Reply by GMatrixGames** [reply]
I did, but I can't find where to find the ownerId for the function. It seems like you don't need it if you're using an admin extension, but since I'm using an embedded app to handle it, I guess I do.

**Reply by aledalgrande** [reply]
You can find the owner (validation) id by running this query https://shopify.dev/docs/api/admin-graphql/2025-01/queries/validations

**Reply by GMatrixGames** [reply]
Thank you so much. I was looking for something under functions rather than a query on the store!

**Comment by aledalgrande** [comment]
Closing as it is implemented.
